### ISSUE TITLE
`ruff` and `libcst`

### ISSUE THREAD

**Issue by kloczek** [astral-sh/ruff#5566]
On building `ruff` I found that it want submodule with libcst
```console
+ /usr/bin/python3 -sBm build -w --no-isolation
* Getting build dependencies for wheel...
* Building wheel...
Running `maturin pep517 build-wheel -i /usr/bin/python3 --compatibility off`
    Updating crates.io index
    Updating git repository `https://github.com/charliermarsh/LibCST`
warning: spurious network error (3 tries remaining): error reading from the zlib stream; class=Zlib (5)
warning: spurious network error (2 tries remaining): error reading from the zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error reading from the zlib stream; class=Zlib (5)
error: failed to get `libcst` as a dependency of package `ruff v0.0.277 (/home/tkloczko/rpmbuild/BUILD/ruff-0.0.277/crates/ruff)`

Caused by:
  failed to load source for dependency `libcst`

Caused by:
  Unable to update https://github.com/charliermarsh/LibCST?rev=80e4c1399f95e5beb532fdd1e209ad2dbb470438#80e4c139

Caused by:
  failed to fetch into: /home/tkloczko/.cargo/git/db/libcst-50aa4195e5d57b22

Caused by:
  error reading from the zlib stream; class=Zlib (5)
💥 maturin failed
  Caused by: Cargo metadata failed. Does your crate compile with `cargo build`?
  Caused by: `cargo metadata` exited with an error:
Error: command ['maturin', 'pep517', 'build-wheel', '-i', '/usr/bin/python3', '--compatibility', 'off'] returned non-zero exit status 1
```
However looks like it is not the same libcst which is listed on pypi (https://github.com/Instagram/LibCST/)

Q: why it is not possible to use `libcst` as installed module and why here is used probably some fork of https://github.com/Instagram/LibCST/? 🤔

**Comment by charliermarsh** [comment]
That looks like a spurious network failure. Are you seeing it repeatedly?

(We have to use a Git dependency as LibCST doesn't publish to crates.io (this is a Rust dependency actually, not a Python dependency), and we fork as their version has some tooling to enable calling LibCST from Python that we don't need for our purposes.)

**Comment by kloczek** [comment]
> That looks like a spurious network failure. Are you seeing it repeatedly?

Nope. I've used just tar ball autogenerated from girt tag which by default is generated out of single repo without submodules.

Disclaimer: I don't know to much about building/maintaining/packaging rust stuff ..

Moment so it is not possible to write `ruff` on top of `libcst` python module? 🤔 
And/or does it mean that if cargo builds exact project it cannot use DSOs generated out of other rust projects and everything is linked statically? 😨 
If that is true it may explain why few python modules which are using `maturin` as pep517 backed produces really huge DSO out of 2-3 KB rust source code 🤔 (sometimes +2MB text code section looking on `size` command output)  -> if it is true it means that rust really .. really sucks 😞

**Comment by charliermarsh** [comment]
Nope, you can't write `ruff` on top of the `libcst` Python module, although I'm not 100% sure that I understand the question and what that would mean. What are you trying to do exactly? Ruff only ships as a binary, not a library.

**Comment by kloczek** [comment]
> Nope, you can't write ruff on top of the libcst Python module

OK so ..

> (We have to use a Git dependency as LibCST doesn't publish to crates.io (this is a Rust dependency actually, not a Python dependency), and we fork as their version has some tooling to enable calling LibCST from Python that we don't need for our purposes.)

Did you try to talk with LibCST  maintainer about publishing on crates.io? 🤔 

Nevertheless official LibCST last version is 1.0.1 and that one on https://github.com/charliermarsh/LibCST is 0.0.1 (looks like typo and probably it should be 1.0.1 as well).

**Comment by kloczek** [comment]
Nevertheless still cannot build `ruff` and `build` with the same errors ..

**Comment by charliermarsh** [comment]
Yeah, there's a note in their README that they want to do it, pull requests are welcome, etc.: https://github.com/Instagram/LibCST/blob/a3f5bf97d631e79c3395a249e15645cfbc225a4d/native/libcst/README.md?plain=1#L10C1-L10C1. It's low enough priority for us that I'm unlikely to do the work required to support it right now, but I will file an issue to make them aware that we would consume it.

> Nevertheless official LibCST last version is 1.0.1 and that one on https://github.com/charliermarsh/LibCST is 0.0.1 (looks like typo and probably it should be 1.0.1 as well).

I think you're confusing the Python library version with the Rust crate which follows different versioning: https://github.com/Instagram/LibCST/blob/main/native/libcst/Cargo.toml.

I'm happy to help with your build issue if you're able to explain what you're trying to do (i.e., why you're trying to build Ruff from source) and the error you're running into (the above looks like a spurious network failure from a Git clone and not related to Ruff, are you seeing it repeatedly?) but right now it's hard to help here based on the info above.

**Comment by kloczek** [comment]
I'm still struggling with complete proper build procedure using isolated from the internet build env.
I' trying to use ruff and https://github.com/Instagram/LibCST/ (1.0.1) tar balls downloaded from git tags.

My understanding is that LibCST tar ball or its part needs to be placed somewhere in ruff source tree.
May I ask for some hints about how to complete source tree assembly procedure using above resources? 🤔

**Comment by zanieb** [comment]
Perhaps https://github.com/astral-sh/ruff/discussions/6680 is a good place for discussion.

I'd recommend looking into `cargo vendor`

**Comment by kloczek** [comment]
@wuch-g2v is my coworker 😋 
We've been trying to complete build procedure however soe far no success 😞

**Comment by kloczek** [comment]
I've been trying to use `cargo vendor` ..
```console
[tkloczko@pers-jacek ruff-0.0.286]$ cargo vendor
    Updating crates.io index
    Updating git repository `https://github.com/Instagram/LibCST.git`
warning: spurious network error (3 tries remaining): error reading from the zlib stream; class=Zlib (5)
warning: spurious network error (2 tries remaining): error reading from the zlib stream; class=Zlib (5)
warning: spurious network error (1 tries remaining): error reading from the zlib stream; class=Zlib (5)
error: failed to sync

Caused by:
  failed to load pkg lockfile

Caused by:
  failed to get `libcst` as a dependency of package `ruff v0.0.286 (/home/tkloczko/rpmbuild/BUILD/ruff-0.0.286/crates/ruff)`

Caused by:
  failed to load source for dependency `libcst`

Caused by:
  Unable to update https://github.com/Instagram/LibCST.git?rev=3cacca1a1029f05707e50703b49fe3dd860aa839#3cacca1a

Caused by:
  failed to fetch into: /home/tkloczko/.cargo/git/db/libcst-b143b6626bde5d59

Caused by:
  error reading from the zlib stream; class=Zlib (5)
```

**Comment by zanieb** [comment]
@kloczek you'll need to vendor the dependencies _with_ internet then take the repository w/ the bundled dependencies and build from that on your offline machine.

**Comment by kloczek** [comment]
Above was executed on system with access to public network.

**Comment by zanieb** [comment]
Hm. Sorry, it runs fine on my machine so unfortunately I cannot provide more guidance. The message sure makes it look like a networking issue though.
